### ISSUE TITLE
Cross compiling from x86-64

### ISSUE THREAD

**Issue by renzenicolai** [0xFelix/linux-mvebu_v5#1]
Hello!

I am trying to build a kernel for my edminiv2 NAS.
Using your package does not work since it only compiles on ARM, thats why I am trying to build a kernel using the armv5 toolchain from the Arch linux website combined with sources from kernel.org and the config you used in this package.

The kernel builds without problems and I can pack it as uImage without problems.

On the NAS the uboot image is detected correctly but after printing "Starting kernel..." the device stops.

Can you supply me with your method (toolchain, commands and if needed patches) for cross compiling the newest linux kernel for this device?

Thank you!

Greetings,
Renze

**Comment by 0xFelix** [comment]
Hello!

You can also compile this package on x86-64 via a little trick using a different config file for makepkg.

This toolchain will work: https://archlinuxarm.org/builder/xtools/x-tools.tar.xz
Take a look at this gist: https://gist.github.com/0xFelix/31e5799c0171e73576656ae068b1ddbc

It is also crucial that you have an u-boot version that is able to boot zImages and pass an initramfs + a device tree blob to the kernel. The stock u-boot is not able to do that.

You could take a look at this repo: https://github.com/0xFelix/u-boot-edminiv2. This can also be compiled with the x-tools linked above, steps are as following:

Set environment variables first:
`export PATH=$PATH:/opt/x-tools/arm-unknown-linux-gnueabi/bin`
`export CROSS_COMPILE=arm-unknown-linux-gnueabi-`

Then issue the following commands:
`make edminiv2_defconfig`
`make all`

After a successful build you will get the files `u-boot-spl.bin` and `u-boot.img`. These need to be written to the edminiv2 NOR flash to the addresses 0xFFFF0000 (u-boot-spl.bin) and 0xFFF90000 (u-boot.img). This can be done via the current u-boot, load them over into ram with `tftp`, then write them to flash with `cp.b`.

Then you need some wizardry to put together an initramfs. Get the latest `ArchLinuxARM-armv5-latest.tar.gz` and extract it, install qemu-arm-static, put `/usr/bin/qemu-arm-static` in the extracted rootfs and chroot into it. Then install the kernel packages created prior and run `mkinitcpio -p linux`. The resulting rootfs can then be put onto your storage.

Fiiinally... start the new u-boot on the edminiv2, load the files `/boot/zImage`, `/boot/dtbs/orion5x-lacie-ethernet-disk-mini-v2.dtb` and `/boot/initramfs-linux.img` into ram and boot the whole thing via `bootz $ZIMAGEADDR $INITRAMADDR $DTBADDR`. 

I don't remember the addresses I used for `tftp` and `bootz` currently but I could look these up over the weekend, just have to dig out my edminiv2... ;-)

Adapting the repo for the newest kernel should also be rather trivial, let me look into this...

Let me know if you need help.

**Comment by renzenicolai** [comment]
Thank you for explaining!

I tried to follow your instructions but I must have messed something up along the way: the device no longer boots the bootloader.

> Marvell>> tftp 500000 u-boot-spl.bin
> Using egiga0 device
> TFTP from server 10.42.1.1; our IP address is 10.42.4.159
> Filename 'u-boot-spl.bin'.
> Load address: 0x500000
> Loading: #
> done
> Bytes transferred = 4556 (11cc hex)
> Marvell>> protect off all
> Un-Protect Flash Bank # 1
> ........... done
> Marvell>> erase all
> Erase Flash Bank # 1 
> ...........
> Erase Flash Bank # 2 Marvell>> cp.b 500000 ffff0000 11cc
> Copy to Flash... done
> Marvell>> tftp 500000 u-boot.img
> Using egiga0 device
> TFTP from server 10.42.1.1; our IP address is 10.42.4.159
> Filename 'u-boot.img'.
> Load address: 0x500000
> Loading: ########################################################
> done
> Bytes transferred = 285552 (45b70 hex)
> Marvell>> cp.b 500000 fff90000 45b70
> Copy to Flash... done

The files I used can be found at "http://rnpl.us/edminiv2_u-boot.zip".

After rebooting the device nothing happens, so something must have gone wrong during flashing or compiling.

Next thing I tried is JTAG using my Buspirate (v3.5).

I connected the bus pirate to J11 as follows: (square pad is pin 1)

J11.3 (TMS) -> cs
J11.4 (CLK) -> clk
J11.5 (TDO) -> miso
J11.6 (TDI) -> mosi

but after starting openocd it gives errors:

> Open On-Chip Debugger 0.9.0 (2016-04-27-23:18)
> Licensed under GNU GPL v2
> For bug reports, read
>   http://openocd.org/doc/doxygen/bugs.html
> Warn : Adapter driver 'buspirate' did not declare which transports it allows; assuming legacy JTAG-only
> Info : only one transport option; autoselect 'jtag'
> srst_only separate srst_gates_jtag srst_open_drain connect_deassert_srst
> trst_and_srst separate srst_gates_jtag trst_push_pull srst_open_drain connect_deassert_srst
> adapter_nsrst_delay: 200
> jtag_ntrst_delay: 200
> Info : Buspirate switched to FAST mode
> Info : Buspirate Interface ready!
> Info : This adapter doesn't support configurable speed
> Error: JTAG scan chain interrogation failed: all zeroes
> Error: Check JTAG interface, timings, target power, etc.
> Error: Trying to use configured scan chain anyway...
> Error: feroceon.cpu: IR capture error; saw 0x00 not 0x01
> Warn : Bypassing JTAG setup events due to errors
> Info : Embedded ICE version 0
> Info : feroceon.cpu: hardware has 1 breakpoint/watchpoint unit
> Error: unexpected Feroceon EICE version signature

Do you see where I went wrong?

**Comment by 0xFelix** [comment]
Did you use the stock u-boot? Maybe I forgot something... It could be that the bone stock u-boot cannot handle the flash properly. You didn't have installed Albert ARIBAUD's version of the stock u-boot by chance? Otherwise the output looks fine, maybe I messed up by pulling the latest changes of the u-boot git. I will check that this evening.

But hope is not lost... I think we can get this working again. I need to check the correct pin assignment later. Did you use my config file for the buspirate? Also very important, which revision of the edminiv2 do you have? Revision C needs some soldering before JTAG works.

**Comment by renzenicolai** [comment]
Yes, I indeed have rev. c of the board. What do I have to solder/fix for the jtag to work? I googled but can not find any documentation about that.

**Comment by 0xFelix** [comment]
Take a look at this picture. You have to solder in a bridge on R169 and a 4.7K ohm resistor on R228.

![edminiv2_revc_jtag](https://cloud.githubusercontent.com/assets/441757/16963678/e3f41ad0-4df7-11e6-83c3-020691f15ffe.jpg)

Concerning the bus pirate JTAG pinout: http://dangerousprototypes.com/docs/Bus_Pirate_I/O_Pin_Descriptions
I have the version with the pinout like in the lower picture, my connections are as following, pin one is the square pad:

J11.1 --> VPU / Green
J11.2 --> GND / Brown
J11.3 --> CS / White
J11.4 --> CLK / Purple
J11.5 --> MISO / Black
J11.6 --> MOSI / Grey
J11.7 --> AUX / Blue
J11.8 --> Unconnected

After soldering and connecting the edminiv2 it should detect the CPU, you should get it into a defined state by issuing `reset init` on the commandline when using my config files. Tell me when you got so far.

**Comment by renzenicolai** [comment]
JTAG works!

I tried to load the u-boot that I had running before using

> load_image /path/to/u-boot-RAM.bin 0x400000
> resume  0x410000

(This is the u-boot that I have running: http://tweaky.aribaud.net/?q=book/export/html/38)

This works and I can use that u-boot to flash and boot it's flash counterpart.

However: when repeating the flashing process as you described before the device again doesn't boot.

Could you send me some known-to-be-working u-boot binaries?

Thank you for all the help!

**Comment by 0xFelix** [comment]
Hey, great news! One suggestion while you have running this old u-boot, note down your MAC address from the environment variables, the new u-boot won't be able to read it out of the EEPROM.

Yeah I noticed that while testing last evening. I somehow messed up by merging with the u-boot master. Currently I am also not able to produce a working binary, I'm at it and will let you know when I got it working.

You're welcome! :-)

**Comment by 0xFelix** [comment]
Hey, got it working again... There is a new branch on my repository called working-state.

I made some binaries for you: [u-boot-working-state-files.zip](https://github.com/0xFelix/linux-mvebu_v5/files/374484/u-boot-working-state-files.zip)

Steps are as following:

In the OpenOCD telnet interface:
`reset init`
`load_image /path/to/u-boot-RAM-0x400000-working-state.bin 0x400000 bin`
`resume 0x400000`

By this time the u-boot CLI should come up on serial. Set up your mac address and the tftp settings in the environment, then do the following  in the u-boot CLI:
`protect off all`
`erase all`
`tftp 0x400000 u-boot-spl-FLASH-0xFFFF0000-working-state.bin`
`cp.b 0x400000 0xFFFF0000 0xBFC`
`tftp 0x400000 u-boot-FLASH-0xFFF90000-working-state.img`
`cp.b 0x400000 0xFFF90000 0x44E50`

After these commands reset the edminiv2 via JTAG or pull the plug. If everything went successful a working u-boot should come up on serial. :-)

Edit:

By the way, concerning the addresses for loading Linux after successfully flashing u-boot.

I got this in my environment for testing:

```
load_all_tftp=run load_kernel; run load_dtb; run load_initramfs
load_dtb=tftp 0x900000 orion5x-lacie-ethernet-disk-mini-v2.dtb
load_initramfs=tftp 0x910000 initramfs-linux.img
load_kernel=tftp 0x400000 zImage
```

Simply run `run load_all_tftp` to load all files, then `bootz 0x400000 0x910000 0x900000`.
